# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSObject.cpp
# Commit: ec72f104edbf
# Full Hash: ec72f104edbfe88e22e2ebd778cc7890c21fd2ce
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 9 - Move proto from ObjectGroup to BaseShape. r=jonco
#   
#   The ObjectGroup still stores the proto as well but we don't really use it anymore.
#   This is simpler than removing the group's proto now because that would result in
#   "empty" groups and the group table then doesn't make sense anymore.
# ==============================================================================

diff -r cad616508f73 -r ec72f104edbf js/src/vm/JSObject.cpp
--- a/js/src/vm/JSObject.cpp	Sat Mar 06 01:05:22 2021 +0000
+++ b/js/src/vm/JSObject.cpp	Sat Mar 06 01:05:22 2021 +0000
@@ -771,9 +771,9 @@
                       ? GetGCKindSlots(gc::GetGCObjectKind(clasp), clasp)
                       : GetGCKindSlots(kind, clasp);
 
-  RootedShape shape(
-      cx, EmptyShape::getInitialShape(cx, clasp, cx->realm(), group->proto(),
-                                      nfixed, objectFlags));
+  RootedShape shape(cx, EmptyShape::getInitialShape(cx, clasp, cx->realm(),
+                                                    group->protoDeprecated(),
+                                                    nfixed, objectFlags));
   if (!shape) {
     return nullptr;
   }
@@ -1202,6 +1202,8 @@
       return false;
     }
 
+    Rooted<BaseShape*> nbase(cx, shape->base());
+
     // Get an in-order list of the shapes in the src object.
     Rooted<ShapeVector> shapes(cx, ShapeVector(cx));
     for (Shape::Range<NoGC> r(src->lastProperty()); !r.empty(); r.popFront()) {
@@ -1213,6 +1215,7 @@
 
     for (Shape* shapeToClone : shapes) {
       Rooted<StackShape> child(cx, StackShape(shapeToClone));
+      child.setBase(nbase);
       shape = cx->zone()->propertyTree().getChild(cx, shape, child);
       if (!shape) {
         return false;
@@ -1818,7 +1821,8 @@
   }
 
   obj->setGroup(newGroup);
-  return true;
+
+  return JSObject::setProtoUnchecked(cx, obj, proto);
 }
 
 /**